crates/credentials/src/pex/presentation_definition.rs

```diff
@@ -0,0 +1,292 @@
+};
+use jsonschema::JSONSchema;
+use rand::Rng;
+use serde::{Deserialize, Serialize};
+use serde_json::{json, to_value, Map, Value};
+
+use crate::pex::PexError;
+
+use super::Result;
+
+/// Represents a DIF Presentation Definition defined [here](https://identity.foundation/presentation-exchange/#presentation-definition).
+/// Presentation Definitions are objects that articulate what proofs a Verifier requires.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct PresentationDefinition {
+    pub id: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub purpose: Option<String>,
+    pub input_descriptors: Vec<InputDescriptor>,
+}
+
+/// Represents a DIF Input Descriptor defined [here](https://identity.foundation/presentation-exchange/#input-descriptor).
+/// Input Descriptors are used to describe the information a Verifier requires of a Holder.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct InputDescriptor {
+    pub id: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub purpose: Option<String>,
+    pub constraints: Constraints,
+}
+
+/// Contains the requirements for a given Input Descriptor.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Constraints {
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    pub fields: Vec<Field>,
+}
+
+/// Contains the requirements for a given field within a proof.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Field {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub id: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    pub path: Vec<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub purpose: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub filter: Option<Filter>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub optional: Option<bool>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub predicate: Option<Optionality>,
+}
+
+/// Type alias for the possible values of the predicate field.
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Optionality {
+    Required,
+    Preferred,
+}
+
+/// A JSON Schema that is applied against the value of a field.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Filter {
+    #[serde(rename = "type", skip_serializing_if = "Option::is_none")]
+    pub r#type: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub pattern: Option<String>,
+    #[serde(rename = "const", skip_serializing_if = "Option::is_none")]
+    pub const_value: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub contains: Option<Box<Filter>>,
+}
+
+fn generate_token() -> String {
```

Comment:
why not uuid? 

:caution: chad generated:
```
[dependencies]
uuid = { version = "1.0", features = ["v4"] } 

extern crate uuid;
use uuid::Uuid;

fn generate_uuid_token() -> String {
    let my_uuid = Uuid::new_v4();
    my_uuid.to_string()
}
```

Reply:
Good call